starlette/middleware/http.py

```diff
@@ -0,0 +1,103 @@
+from typing import AsyncGenerator, Callable, Optional, Union
+
+from .._compat import aclosing
+from ..datastructures import MutableHeaders
+from ..requests import HTTPConnection
+from ..responses import Response
+from ..types import ASGIApp, Message, Receive, Scope, Send
+
+_DispatchFlow = Union[
+    # Default case:
+    # response = yield
+    AsyncGenerator[None, Response],
+    # Early response and/or error handling:
+    # if condition:
+    #     yield Response(...)
+    #     return
+    # try:
+    #     response = yield None
+    # except Exception:
+    #     yield Response(...)
+    # else:
+    #    ...
+    AsyncGenerator[Optional[Response], Response],
+]
+
+
+class HTTPMiddleware:
+    def __init__(
+        self,
+        app: ASGIApp,
+        dispatch: Optional[Callable[[HTTPConnection], _DispatchFlow]] = None,
+    ) -> None:
+        if dispatch is None:
+            dispatch = self.dispatch
+
+        self.app = app
+        self._dispatch_func = dispatch
+
+    def dispatch(self, conn: HTTPConnection) -> _DispatchFlow:
+        raise NotImplementedError  # pragma: no cover
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if scope["type"] != "http":
+            await self.app(scope, receive, send)
+            return
+
+        conn = HTTPConnection(scope)
+
+        async with aclosing(self._dispatch_func(conn)) as flow:
+            # Kick the flow until the first `yield`.
+            # Might respond early before we call into the app.
+            maybe_early_response = await flow.__anext__()
+
+            if maybe_early_response is not None:
+                await maybe_early_response(scope, receive, send)
+                return
+
+            response_started: set = set()
+
+            async def wrapped_send(message: Message) -> None:
+                if message["type"] == "http.response.start":
+                    response_started.add(True)
+
+                    response = Response(status_code=message["status"])
+                    response.raw_headers.clear()
+
+                    try:
+                        await flow.asend(response)
+                    except StopAsyncIteration:
+                        pass
+                    else:
+                        raise RuntimeError("dispatch() should yield exactly once")
+
+                    headers = MutableHeaders(raw=message["headers"])
+                    headers.update(response.headers)
```

Comment:
```suggestion
                    headers = MutableHeaders(raw=message["headers"])
                    headers.update(response.headers)
                    message["headers"] = headers.raw
```

I'm guessing `MutableHeaders` mutates the list it receives, but it's not super obvious and I'm not sure we should rely on that behavior.

Reply:
This is a pattern we use in a number of places in the codebase already, eg. here: https://github.com/encode/starlette/blob/5b56e7dc429de946be8c782e0ae69bcc5d5b5585/starlette/middleware/cors.py#L157-L158

`MutableHeaders` takes care of deduplicating and normalization for us, which we wouldn't get with eg. `message["headers"].extend(response.raw_headers)`.

Reply:
I get that, the part I don't like is that we don't explicitly assign `message["headers"] = MutableHeaders.raw`. I don't like relying on `MutableHeaders` modifying the list it gets passed in place. This is counterintuitive: `a = [1]; list(a).pop()` doesn't modify `a`.  So even if in practice it is unnecessary I think it is worth explicitly assigning to be clear about what is happening.

Reply:
Oh, I see.